Issue #735 Extend HasGrenadePocket() and HasAmmoPocket()

Musashi1584 opened the pull request:
Closes #735

Xymanek commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc:
```diff
@@ -3758,28 +3758,172 @@ function bool MeetsAbilityPrerequisites(name AbilityName)
+	// End Issue #735 (1/2)
+}
+
+/// HL-Docs: feature:OverrideHasAmmorPocket; issue:735; tags:itemslots,strategy
```
OverrideHasAmmo**r**Pocket

Musashi1584 replied:
fixed

Xymanek commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc:
````diff
@@ -3758,28 +3758,172 @@ function bool MeetsAbilityPrerequisites(name AbilityName)
+/// Finally the event AbilityUnlocksAmmoPocket is triggered that allows mods to override the final result
+///
+/// ```unrealscript
+/// EventID: OverrideHasGrenadePocket
````
Copy paste slip up

Musashi1584 replied:
fixed

Xymanek commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc:
```diff
@@ -3758,28 +3758,172 @@ function bool MeetsAbilityPrerequisites(name AbilityName)
+	Tuple.Data[0].kind = XComLWTVBool;
+	Tuple.Data[0].b = bHasAmmoPocket;
+
+	`XEVENTMGR.TriggerEvent('OverrideHasGrenadePocket', Tuple, self, none);
```
Copy paste slip up

Musashi1584 replied:
fixed

Musashi1584 commented:
> Are we sure that making it by default pull from item abilities array is backwards compatible?

I see no scenarion why it shouldnt be. If someone manually added an ability from an item to the config array the only difference is that it works now as he expected it.

pledbrook commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc:
```diff
@@ -3758,28 +3758,172 @@ function bool MeetsAbilityPrerequisites(name AbilityName)
 {
 	local name CheckAbility;
 
+	// Variables for Issue #735 (1/2)
```
Would you mind adding an `// End` line or other boundary marker so we know how many of the following variables are associated with issue #735? Don't want to have to look at the associated commit just to find out that information.

pledbrook commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc:
```diff
@@ -3758,28 +3758,172 @@ function bool MeetsAbilityPrerequisites(name AbilityName)
+	if (!bHasGrenadePocket)
+	{
+		CurrentInventory = GetAllInventoryItems();
+		foreach CurrentInventory(InventoryItem)
```
Might a
```
for (idx = 0; i < CurrentInventory.Length && !bHasGrenadePocket)
```
be slightly more comprehensible than using `break`s? Not a deal breaker, it's just the following code is not super easy to grok.

pledbrook replied:
Same comment applies to `HasAmmoPocket()`. Perhaps there isn't a nice way to do this without factoring out some additional functions. Go with whatever you prefer.

MrNiceUK replied:
Depends what you're used to, keeping the for condition simple (or entirely implicit with foreach) then breaking when you find an object of interest is a pattern I've seen often, including in Firaxis X2 code, for it to grok fine

MrNiceUK replied:
Depends what you're used to, keeping the for condition simple (or entirely implicit with foreach) then breaking when you find an object of interest is a pattern I've seen often, including in Firaxis X2 code, for it to grok fine

pledbrook commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc:
```diff
@@ -3758,28 +3758,172 @@ function bool MeetsAbilityPrerequisites(name AbilityName)
 {
 	local name CheckAbility;
 
+	// Variables for Issue #735 (2/2)
```
Same suggestion for an `// End Issue` line or similar.

Musashi1584 replied:
Below should only come variables of the issue. If another issue adds variables they should start with a new comment

pledbrook replied:
Let me put it another way: if I come across this code 3 months from now, could I be confident at first glance that those variables are all linked to that one issue? No, because I don't know who wrote the code or what conventions they were using at the time. Anyway, if you'd prefer to leave it as is, no worries.

Musashi1584 commented:
@pledbrook 
I refactored the code a bit and added the expanded checks and the event `HasExtraUtilitySlot` as well. Can you do another review when you have some time to spare?

pledbrook commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc:
```diff
@@ -3758,50 +3758,221 @@ function bool MeetsAbilityPrerequisites(name AbilityName)
+/// 	  inout bool bHasGrenadePocket
+///     ]
+/// }
+/// EventSource: XComGameState_Unit
```
I would add a doc line for `NewGameState`, i.e. `NewGameState: yes`, or `NewGameState: no`. This is just the let mod authors know whether the event is passed with a `NewGameState` instance or not (if they use `ELD_Immediate`) that they can use to make more game state changes.

pledbrook commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc:
```diff
@@ -3758,50 +3758,221 @@ function bool MeetsAbilityPrerequisites(name AbilityName)
+	if (!bHasGrenadePocket)
+	{
+		bHasGrenadePocket = HasAnyOfTheAbilitiesFromCharacterTemplate(class'X2AbilityTemplateManager'.default.AbilityUnlocksGrenadePocket);
+	}
```
Would it make sense to have an additional public function that will check whether a unit has any (and maybe one for all) abilities from any source? I was thinking that something like Quick End Turn would benefit from a function that could be used to determine whether a unit has a Steady Weapon ability, regardless of where it came from.

MrNiceUK replied:
Just to note, while in Tactical `HasSoldierAbility()` already does that (if the second parameter is true, but that's the default value for it anyway). As the comment on it says, that uses `FindAbility()` behind the scenes, which itself uses the `Abilities` array, which is only populated during tactical. Quick reminder, the whole issue here is that the `Has...Pocket()` methods have to work from loadout/armory too.

pledbrook replied:
OK, good to know, thanks. I still think `HasAny/AllTheAbilitiesFromAnySource()` functions would be useful, if only to be reused from the three places in this pull request.

pledbrook commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc:
```diff
@@ -3758,50 +3758,221 @@ function bool MeetsAbilityPrerequisites(name AbilityName)
 
+	`XEVENTMGR.TriggerEvent('OverrideHasGrenadePocket', Tuple, self, none);
+	
+	return Tuple.Data[0].b;
```
I would prefer that this block of code were put into a separate function to isolate the tuple boilerplate from the real meat of the function. The tuple stuff is just a distraction and doesn't help readers understand what the `HasGrenadePocket()` is doing.

With this change and the one suggested above, the function would look like:
```
function bool HasGrenadePocket()
{
    local bool bHasGrenadePocket;

    bHasGrenadePocket = HasAnyOfTheAbilitiesFromAnySource(class'X2AbilityTemplateManager'.default.AbilityUnlocksGrenadePocket);
    return TriggerOverrideHasGrenadePocket(bHasGrenadePocket);
}
```
In fact, it could be made a one liner, but that may be a bit too over the top.

pledbrook replied:
In fact, the tuple boilerplate is the same for all three cases. The only difference is the event name. So you really only need one `TriggerHasPocketOfTypeEvent(name EventName)` function that can be called from all three places.

pledbrook commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc:
```diff
@@ -3758,50 +3758,221 @@ function bool MeetsAbilityPrerequisites(name AbilityName)
+/// 	  inout bool bHasAmmoPocket
+///     ]
+/// }
+/// EventSource: XComGameState_Unit
```
Same comments for HasGrenadePocket apply to HasAmmoPocket of course.

pledbrook commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc:
```diff
@@ -3758,50 +3758,221 @@ function bool MeetsAbilityPrerequisites(name AbilityName)
+/// 	  inout bool bHasExtraUtilitySlot
+///     ]
+/// }
+/// Even
```
Typo or copy/paste error? Assume this should be 3 backticks to close off the code block.

Musashi1584 replied:
I think my latest changes address all change requests. Please check and resolve if everything looks good.

Iridar commented:
Not sure if it's of much of a concern, but `HasAnyOfTheAbilitiesFromInventory` does not check abilities attached to weapon upgrades.
